File: 55_02_BalancedBinaryTree/BalancedBinaryTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    res = True

    def IsBalanced_Solution(self, pRoot):
        # write code here
        self.isBalancedHelper(pRoot)
        return self.res

    def isBalancedHelper(self, root):
        if not root:
            return 0
        if not self.res:
            return 1
        left = 1 + self.isBalancedHelper(root.left)
        right = 1 + self.isBalancedHelper(root.right)
        if abs(left - right) > 1:
            self.res = False
        return max(left, right)

File: 55_02_BalancedBinaryTree/BalancedBinaryTree.cpp
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

class Solution
{
public:
    bool IsBalanced_Solution(TreeNode *pRoot)
    {
        int depth = 0;
        return IsBalanced(pRoot, depth);
    }

    bool IsBalanced(TreeNode *root, int &depth)
    {
        if (root == nullptr)
        {
            depth = 0;
            return true;
        }
        int left, right;
        if (IsBalanced(root->left, left) && IsBalanced(root->right, right))
        {
            int diff = left - right;
            if (diff <= 1 && diff >= -1)
            {
                depth = 1 + (left > right ? left : right);
                return true;
            }
        }
        return false;
    }
};